pennylane/resource/first_quantization.py

```diff
@@ -589,3 +651,354 @@ def qubit_cost(n, eta, omega, error, br=7, charge=0):
+        n_steps = 30  # maximum number of amplitude amplification steps
+        for i in reversed(range(n_steps)):
+            probability = (np.sin((2 * i + 1) * np.arcsin(np.sqrt(p_nu)))) ** 2
+            if probability > p_th:
+                aa_steps = i
+                p_nu_amp = probability
+
+        # lambda_u and lambda_v are taken from Eq. (25) of PRX Quantum 2, 040332 (2021)
+        lambda_u = 4 * np.pi * eta * l_z * lambda_nu / omega
+        lambda_v = 2 * np.pi * eta * (eta - 1) * lambda_nu / omega
+
+        # lambda_u_1 and lambda_v_1 are taken from Eq. (124) of PRX Quantum 2, 040332 (2021)
+        lambda_u_1 = lambda_u * lambda_nu_1 / lambda_nu
+        lambda_v_1 = lambda_v * lambda_nu_1 / lambda_nu
+
+        b_mat = np.matrix(recip_vectors)
+        abs_sum = np.abs(b_mat @ b_mat.T).flatten().sum()
+
+        # taken from Appendix I.1 of arXiv:2302.07981 (2023)
+        if orthogonal:
+            lambda_t_p = abs_sum * eta * 2 ** (2 * n_p - 2) / 4
+        else:
+            lambda_t_p = abs_sum * eta * 2 ** (2 * n_p - 2) / 2
+
+        # taken from Eq. (63) of PRX Quantum 2, 040332 (2021)
+        p_eq = (
+            FirstQuantization.success_prob(3 * eta + 2 * charge, br)
+            * FirstQuantization.success_prob(eta, br) ** 2
+        )
+
+        # final lambda value is computed from case checking derived in Section E and Appendix D of
+        # PRX Quantum 2, 040332 (2021)
+        if p_nu * lambda_t_p >= (1 - p_nu) * (lambda_u_1 + lambda_v_1):
+            return 0.0, 0.0
+        if p_nu_amp * lambda_t_p >= (1 - p_nu_amp) * (lambda_u_1 + lambda_v_1):
+            return (lambda_t_p + lambda_u_1 + lambda_v_1) / p_eq, aa_steps
+        return ((lambda_u_1 + lambda_v_1 / (1 - 1 / eta)) / p_nu_amp) / p_eq, aa_steps
+
+    @staticmethod
+    def _qubit_cost_noncubic(n, eta, error, br, charge, vectors):
+        r"""Return the number of logical qubits needed to implement the first quantization
+        algorithm for non-cubic systems.
+
+        Args:
+            n (int): number of plane waves
+            eta (int): number of electrons
+            error (float): target error in the algorithm
+            br (int): number of bits for ancilla qubit rotation
+            charge (int): total electric charge of the system
+            vectors (array[float]): lattice vectors
+
+        Returns:
+            int: number of logical qubits needed to implement the first quantization algorithm
+        """
+
+        lambda_total, _ = FirstQuantization._norm_noncubic(
+            n,
+            eta,
+            error,
+            br,
+            charge,
+            vectors,
+        )
+
+        omega = np.abs(np.sum((np.cross(vectors[0], vectors[1]) * vectors[2])))
+
+        recip_vectors = (  # taken from Eq. (35) of arXiv:2302.07981 (2023)
+            2
+            * np.pi
+            / omega
+            * np.array([np.cross(vectors[i], vectors[j]) for i, j in [(1, 2), (2, 0), (0, 1)]])
+        )
```

Comment:
It is not very time consuming and we want the functions to be self contained. Otherwise, if a user wants to use the functions individually, they have to input all these parameters.